test/cpp/api/inference_mode.cpp

```diff
@@ -0,0 +1,391 @@
+#include <torch/script.h>
+#include <gtest/gtest.h>
+#include <test/cpp/api/support.h>
+
+using namespace torch::autograd;
+using namespace torch::test;
+
+namespace {
+  torch::Tensor functional_op(torch::Tensor& x) {
+    return x + x;
+  }
+
+  void inplace_op(torch::Tensor& x) {
+    x.add_(1);
+  }
+
+  torch::Tensor view_op(torch::Tensor& x) {
+    return x.view({2, 3});
+  }
+}
+
+/*
+  Only the following combos of Autograd & InplaceOrView keys on tensors are valid:
+    - Autograd=true, InplaceOrView=true (normal tensor)
+    - Autograd=false, InplaceOrView=false (inference tensor)
+  Tensors created in InferenceMode are mostly inference tensors. There's an exception
+  which listed below, even though these tensors are created in InferenceMode, they're
+  normal tensors:
+    - tensors which are view of normal tensors.
+  Note output of inplace ops on normal tensors(which is the input itself).
+*/
+bool is_inference_tensor(torch::Tensor& x) {
+  c10::DispatchKeySet ks = x.key_set();
+  bool has_Autograd = ks.has(c10::DispatchKey::AutogradCPU);
+  bool has_InplaceOrView = ks.has(c10::DispatchKey::InplaceOrView);
+  // They must be either both true or false.
+  bool is_inference_tensor = !has_Autograd && !has_InplaceOrView && x.is_leaf();
+  return is_inference_tensor;
+}
+
+void assert_tensor_creation_meta(torch::Tensor& x, torch::autograd::CreationMeta creation_meta) {
+  ASSERT_EQ(static_cast<torch::autograd::DifferentiableViewMeta*>(x.unsafeGetTensorImpl()->autograd_meta())->get_creation_meta(), creation_meta);
+}
+
+void assert_TLS_states(bool inference_mode) {
+  ASSERT_EQ(InferenceMode::is_enabled(), inference_mode);
+  ASSERT_EQ(c10::impl::tls_is_dispatch_keyset_excluded(c10::autograd_dispatch_keyset), inference_mode);
+  ASSERT_EQ(c10::impl::tls_is_dispatch_key_included(c10::DispatchKey::InplaceOrView), !inference_mode);
+}
+
+TEST(InferenceModeTest, TestTLSState) {
+  assert_TLS_states(false);
+  {
+    InferenceMode guard;
+    assert_TLS_states(true);
+    {
+      InferenceMode guard(false);
+      assert_TLS_states(false);
+    }
+    assert_TLS_states(true);
+  }
+  assert_TLS_states(false);
+}
+
+TEST(InferenceModeTest, TestInferenceTensorCreation) {
+  {
+    InferenceMode guard;
+    // New tensor created through constructors are inference tensors.
+    torch::Tensor c = torch::ones({1, 2, 3});
+    ASSERT_FALSE(c.requires_grad());
+    ASSERT_TRUE(is_inference_tensor(c));
+
+    torch::Tensor tmp = torch::ones({1, 2, 3}).set_requires_grad(true);
+    ASSERT_TRUE(tmp.requires_grad()); // requires_grad=true on inference tensor is no-op.
```

Comment:
not sure what that comment means.
setting that field is not a noop as you actually properly set it.
But it will be ignored when performing computations.